bitseq/sequence.go

```diff
@@ -297,75 +296,91 @@ func (h *Handle) CheckConsistency() error {
 	}
 }
 
+// Refresh updates current datastore view to latest
+func (h *Handle) Refresh() error {
+	if h.store != nil {
+		if err := h.store.GetObject(datastore.Key(h.Key()...), h); err != nil && err != datastore.ErrKeyNotFound {
+			return err
+		}
+	}
+	return nil
+}
+
 // set/reset the bit
 func (h *Handle) set(ordinal, start, end uint64, any bool, release bool) (uint64, error) {
 	var (
 		bitPos  uint64
 		bytePos uint64
 		ret     uint64
 		err     error
+		mod     bool
 	)
 
+	if h.store == nil {
+		h.serial <- struct{}{}
+		defer func() { <-h.serial }()
+	}
+
 	for {
-		var store datastore.DataStore
-		h.Lock()
-		store = h.store
-		h.Unlock()
-		if store != nil {
-			if err := store.GetObject(datastore.Key(h.Key()...), h); err != nil && err != datastore.ErrKeyNotFound {
-				return ret, err
-			}
+		if err := h.Refresh(); err != nil {
+			return invalidPos, err
 		}
 
-		h.Lock()
+		// Create a private copy of h and work on it
+		nh := h.getCopy()
+
 		// Get position if available
-		if release {
+		if release || !any {
 			bytePos, bitPos = ordinalToPos(ordinal)
+			ret = ordinal
 		} else {
-			if any {
-				bytePos, bitPos, err = getFirstAvailable(h.head, start)
-				ret = posToOrdinal(bytePos, bitPos)
-				if end < ret {
-					err = ErrNoBitAvailable
-				}
-			} else {
-				bytePos, bitPos, err = checkIfAvailable(h.head, ordinal)
-				ret = ordinal
+			bytePos, bitPos, err = getFirstAvailable(nh.head, start)
+			if err != nil {
+				return invalidPos, err
+			}
+			ret = posToOrdinal(bytePos, bitPos)
+			if ret > end {
+				return invalidPos, ErrNoBitAvailable
 			}
-		}
-		if err != nil {
-			h.Unlock()
-			return ret, err
 		}
 
-		// Create a private copy of h and work on it
-		nh := h.getCopy()
-		h.Unlock()
-
-		nh.head = pushReservation(bytePos, bitPos, nh.head, release)
-		if release {
-			nh.unselected++
+		if nh.head, mod = pushReservation(bytePos, bitPos, nh.head, release); mod {
+			if release {
+				nh.unselected++
+			} else {
+				nh.unselected--
+			}
 		} else {
-			nh.unselected--
+			if !release && !any {
+				return invalidPos, ErrBitAllocated
+			} else if release {
+				logrus.Warnf("Redundant release on %q during (release: %t, any: %t), (ordinal: %d, ret: %d)",
+					nh.id, release, any, ordinal, ret)
+				return ret, nil
+			} else {
+				// This should not happen, we should panic
+				return invalidPos, fmt.Errorf("Unexpected redundant reserve on %q during (release: %t, any: %t), (ordinal: %d, ret: %d)",
+					nh.id, release, any, ordinal, ret)
+			}
 		}
 
 		// Attempt to write private copy to store
-		if err := nh.writeToStore(); err != nil {
-			if _, ok := err.(types.RetryError); !ok {
-				return ret, fmt.Errorf("internal failure while setting the bit: %v", err)
-			}
-			// Retry
-			continue
+		if err = nh.writeToStore(); err == nil {
+			h.Lock()
```

Comment:
nh.writeStore() takes the datastore lock. So in this scenario we take in h.lock followed by ds.lock (part of putObjectAtomic).

Now at the start of the set function we take the ds.lock (part of getObject) and then take h.lock() (also part of getObject ->> cache.get() ) So this will cause 2 go routines to go into a deadlock where go1 will take ds.lock and wait on h.lock and go2 will take in h.lock and wait on ds.lock